workspace: Move checkWorkspace

frznlogic opened the pull request:
This commit moves the checkWorkspace method from the SoftwareContainer to the
ContainerUtilityInterface and the usage into the SoftwareContainerAgent class
as the workspace is connected to the agent, not each container. A new check
was added to the SoftwareContainer for the containerRoot (previously
workspace).

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

frznlogic commented:
@breakreturn could you make sure you are ok with the changes to the ContainerUtilityInterface. I'm not sure if this is the right way of doing this and thought i'd push it early so you can have some input.

breakreturn commented on agent/src/containerutilityinterface.h:
```diff
@@ -61,15 +65,24 @@ class ContainerUtilityInterfaceError : public SoftwareContainerError
 
-class ContainerUtilityInterface {
+class ContainerUtilityInterface :
+    private FileToolkitWithUndo {
```
Put the brace on the next line or put the inheritance stuff on the same line please.

frznlogic replied:
Fixed

kursatkobya commented on agent/component-test/softwarecontaineragent_componenttest.cpp:
```diff
@@ -105,7 +105,7 @@ class SoftwareContainerAgentTest: public ::testing::Test
         try {
             std::shared_ptr<SoftwareContainerFactory> factory = std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface(config));
```
to be concise could you use somethin like that instead 
`auto utility = std::make_shared<ContainerUtilityInterface>(config);`

frznlogic replied:
Fixed for both instances.

tobsan commented on agent/component-test/softwarecontaineragent_componenttest.cpp:
```diff
@@ -105,7 +105,7 @@ class SoftwareContainerAgentTest: public ::testing::Test
         try {
             std::shared_ptr<SoftwareContainerFactory> factory = std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface(config));
```
I would prefer you only pass the one value you need from the config. If we need more values later on we can change it to take the whole config as an argument.

frznlogic replied:
I tend to disagree. Or rather, I'm ambivalent as I can see a need for the data inside the class as I interpret the use of it. Would you be terribly upset at keeping it?

kursatkobya commented on agent/src/containerutilityinterface.cpp:
```diff
@@ -70,4 +78,21 @@ void ContainerUtilityInterface::removeOldContainers()
+    const std::string rootDir = m_config->getStringValue("SoftwareContainer", "shared-mounts-dir");
+    if (!isDirectory(rootDir)) {
+        log_debug() << "Container root " << rootDir << " does not exist, trying to create";
+        std::unique_ptr<CreateDir> createDirInstance = std::unique_ptr<CreateDir>(new CreateDir());
```
and maybe initiate it like 
`std::unique_ptr<CreateDir> createDirInstance{new CreateDir()}`

frznlogic replied:
Not sure I understand?

frznlogic replied:
Ok, done!

tobsan commented on libsoftwarecontainer/src/container.cpp:
```diff
@@ -190,9 +190,9 @@ bool Container::create()
-        const std::string rootFSPathWork  = m_containerRoot + m_id + "/rootfs-work";
+        const std::string rootFSPathLower = m_containerRoot + "/rootfs-lower";
+        const std::string rootFSPathUpper = m_containerRoot + "/rootfs-upper";
+        const std::string rootFSPathWork  = m_containerRoot + "/rootfs-work";
```
This seems to be unrelated with the checkContainerRoot stuff. Please split it into its own commit.

frznlogic replied:
It's related, the path available in SoftwareContainer is now "proper", ie it contains the m_id also.

kursatkobya commented on agent/src/main.cpp:
```diff
@@ -219,7 +219,7 @@ int main(int argc, char **argv)
             std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
         std::shared_ptr<ContainerUtilityInterface> utility =
-            std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface(config));
```
again using make_shared would make it more concise and readable

frznlogic commented:
I will split the commits apart a bit tomorrow.

Commit:
createdir: Add pragma once

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

Commit:
workspace: Move checkWorkspace

This commit moves the checkWorkspace method from the SoftwareContainer to the
ContainerUtilityInterface and the usage into the SoftwareContainerAgent class
as the workspace is connected to the agent.

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

Commit:
writebuffer: Add tmpfs support

This commit adds support for tmpfs mounting of the container temp
directory if the write buffer is enabled.

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

tobsan commented on agent/component-test/softwarecontaineragent_componenttest.cpp:
```diff
@@ -104,8 +104,8 @@ class SoftwareContainerAgentTest: public ::testing::Test
-            std::shared_ptr<SoftwareContainerFactory> factory = std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-            std::shared_ptr<ContainerUtilityInterface> utility = std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+            auto factory = std::make_shared<SoftwareContainerFactory> ();
+            auto utility = std::make_shared<ContainerUtilityInterface> (config);
```
Remove the spaces before opening parenthesis.

tobsan commented on agent/src/containerutilityinterface.cpp:
```diff
@@ -70,4 +78,21 @@ void ContainerUtilityInterface::removeOldContainers()
+    const std::string rootDir = m_config->getStringValue("SoftwareContainer", "shared-mounts-dir");
+    if (!isDirectory(rootDir)) {
+        log_debug() << "Container root " << rootDir << " does not exist, trying to create";
+        std::unique_ptr<CreateDir> createDirInstance{new CreateDir()};
```
Why the { syntax instead of ( ?

frznlogic replied:
Copy paste from Kursats comment before, it's just a syntax difference in this case I think. It does look nicer than the original way of writing it however.

kursatkobya replied:
it is also safer in small aspect. one can mistakenly use () and create a function instead of running constructor of variable i thought.

tobsan commented on agent/src/containerutilityinterface.h:
```diff
@@ -61,15 +65,24 @@ class ContainerUtilityInterfaceError : public SoftwareContainerError
 public:
-    virtual ~ContainerUtilityInterface() {}
+    ContainerUtilityInterface(std::shared_ptr<Config> config);
+    ~ContainerUtilityInterface() {}
```
The destructor should be virtual if we want to be able to mock this interface.

frznlogic replied:
Ah, ok... Changing it back

tobsan commented on common/filetoolkitwithundo.cpp:
```diff
@@ -174,6 +174,35 @@ bool FileToolkitWithUndo::syncOverlayMount(const std::string &lower,
+        return false;
+    }
+
+    std::string mountoptions = logging::StringBuilder() << "size=" << maxSize;
```
This can be done with std::to_string as well I think. No need to change now, just a thought.

frznlogic replied:
Thanks;)

tobsan commented on libsoftwarecontainer/src/softwarecontainer.cpp:
```diff
@@ -61,18 +61,25 @@ SoftwareContainer::SoftwareContainer(const ContainerID id,
+    m_containerRoot = buildPath(m_config->sharedMountsDir(), "SC-" + std::to_string(id));
+    checkContainerRoot(m_containerRoot);
+    if (m_config->enableWriteBuffer()) {
+        tmpfsMount(m_containerRoot, 100485760);
```
Magic number.

frznlogic replied:
The next story is about removing the magic number ;).

tobsan replied:
I'm a bit reluctant on merging things that are super static. Can't we at least have a variable holding the number or something so that it is clear that it shouldn't be static?

frznlogic replied:
Fair enough. Fixed in new updates ;).

Commit:
fixes: General syntax fixes

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

Commit:
cookbook: bump

Adds psutil to the pytest dependencies

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>

breakreturn commented on agent/src/main.cpp:
```diff
@@ -216,10 +216,8 @@ int main(int argc, char **argv)
-            std::shared_ptr<SoftwareContainerFactory> (new SoftwareContainerFactory());
-        std::shared_ptr<ContainerUtilityInterface> utility =
-            std::shared_ptr<ContainerUtilityInterface> (new ContainerUtilityInterface());
+        auto factory = std::make_shared<SoftwareContainerFactory> ();
```
There shouldn't be spaces before the parenthesis.

breakreturn commented on agent/unit-test/softwarecontaineragent_unittest.cpp:
```diff
@@ -152,7 +154,7 @@ class SoftwareContainerAgentTest: public ::testing::Test
         testContainerInterface = std::shared_ptr<::testing::NiceMock<TestContainerInterface>> (new ::testing::NiceMock<TestContainerInterface>());
         factory = std::shared_ptr<SoftwareContainerFactory> (new TestFactory(testContainerInterface));
-        containerUtility = std::shared_ptr<::testing::NiceMock<MockUtility>> (new ::testing::NiceMock<MockUtility>());
+        containerUtility = std::shared_ptr<::testing::NiceMock<MockUtility>> (new ::testing::NiceMock<MockUtility>(config));
```
Remove the space before parenthesis.

breakreturn commented on libsoftwarecontainer/src/softwarecontainer.cpp:
```diff
@@ -61,18 +61,26 @@ SoftwareContainer::SoftwareContainer(const ContainerID id,
                       m_config->containerShutdownTimeout()));
 
+
+
```
Why all the whitespace here?

frznlogic replied:
Because I like it :D (Fixed)

breakreturn commented on servicetest/filesystem/test_filesystem.py:
```diff
@@ -202,3 +203,56 @@ def test_bindmount_sockets(self, flag):
+    @pytest.mark.parametrize("flag", [True, False])
+    def test_tmpfs_writebuffer_size(self, flag):
+        """ The SoftwareContainer will create a tmpfs in the location of the
+        containers temporary directory location if the write buffer is enabled.
```
Formatting of these comment are not as we usually do it, please have a look through the code for style and compare to the other modules in the service test tree.

breakreturn commented on servicetest/filesystem/test_filesystem.py:
```diff
@@ -202,3 +203,56 @@ def test_bindmount_sockets(self, flag):
+
+        ca = Container()
+
+        if flag is True:
```
I'm not a fan of using parametrization like this. What's the benefit compared to having two tests? The code is mostly not reused.

Commit:
Fix: syntax, indentation and so forth

Signed-off-by: Oscar Andreasson <oscar.andreasson@pelagicore.com>
